api/src/span_builder.rs

```diff
@@ -0,0 +1,36 @@
+use actix_web::{
+    body::MessageBody,
+    dev::{ServiceRequest, ServiceResponse},
+    Error,
+};
+use tracing::Span;
+use tracing_actix_web::{DefaultRootSpanBuilder, RootSpanBuilder};
+
+/// The `RootSpanBuilder` implementation for the API service.
+/// It extracts the project ref from the `tenant_id` header
+/// and sets it as a field in the root span. If the header is not
+/// present, it sets the field to `Empty` to allow handlers
+/// to set it later.
+pub struct ApiRootSpanBuilder;
+
+impl RootSpanBuilder for ApiRootSpanBuilder {
+    fn on_request_start(request: &ServiceRequest) -> Span {
+        let project = request.headers().get("tenant_id");
+        match project {
+            Some(project) => {
+                // We convert lossily to a string to be able to read at least part of the
+                // project ref in case of invalid UTF-8. This is useful for debugging.
+                // This is anyways an extreme edge case, as the project ref is
+                // generated by the system and should be valid UTF-8.
```

Comment:
Is it in the spec of our project ref that it is a UTF-8?

Reply:
A project id looks something like this: `abcdpblmofovptfcrxyz` i.e. a 20 alphabets long id. So should be a valid UTF-8 always.

Reply:
Yeah, I thought we want to be more restrictive and allow only ASCII characters (which the ref seems to be composed of).

Reply:
A more restrictive type may be not be useful here as this is mainly for tracing, but I agree that we could have such a type elsewhere in the code.